### PR TITLE
Validate document structure

### PR DESCRIPTION
**Pull request by millicentachieng**
Validate document outline based on api page type.

Document outline config is in the below format:

```
{
  "apiPageType": [
    {
      "title": "",
      "level": 1,
      "required": true,
      "headers": [
        {
          "title": "Permissions",
          "level": 2,
          "required": true
        },
        {
          "title": "HTTP request",
          "level": 2,
          "required": true
        },
        {
          "title": "Path parameters",
          "level": 2,
          "required": false
        },
        {
          "title": "Function parameters",
          "level": 2,
          "required": false
        },
        {
          "condition": "OR",
          "arguments": [
            {
              "title": "Query parameters",
              "level": 2,
              "required": false,
              "headers": [
                {
                  "title": "OData query parameters",
                  "level": 3,
                  "required": false
                }
              ]
            },
            {
              "title": "Optional query parameters",
              "level": 2,
              "required": false
            }
          ]
        },
        {
          "title": "Request headers",
          "level": 2,
          "required": true
        },
        {
          "title": "Request body",
          "level": 2,
          "required": true
        },
        {
          "title": "Response",
          "level": 2,
          "required": true
        },
        {
          "title": "Examples",
          "level": 2,
          "required": true,
          "headers": [
            {
              "condition": "OR",
              "arguments": [
                {
                  "title": "Example *",
                  "level": 3,
                  "allowMultiple": true,
                  "required": true,
                  "headers": [
                    {
                      "title": "Request",
                      "level": 4,
                      "required": true
                    },
                    {
                      "title": "Response",
                      "level": 4,
                      "required": true
                    }
                  ]
                },
                {
                  "condition": "AND",
                  "arguments": [
                    {
                      "title": "Request",
                      "level": 3,
                      "required": true
                    },
                    {
                      "title": "Response",
                      "level": 3,
                      "required": true
                    }
                  ]
                }
              ]
            }
          ]
        },
        {
          "title": "See also",
          "level": 2,
          "required": false
        }
      ]
    }
  ],
  "resourcePageType": [
    {
      "title": "* resource type",
      "level": 1,
      "required": true,
      "headers": [
        {
          "title": "Methods",
          "level": 2,
          "required": false
        },
        {
          "title": "Properties",
          "level": 2,
          "required": true
        },
        {
          "title": "Relationships",
          "level": 2,
          "required": false
        },
        {
          "title": "JSON representation",
          "level": 2,
          "required": true
        },
        {
          "title": "See also",
          "level": 2,
          "required": false
        }
      ]
    }
  ],
  "enumPageType": [
    {
      "title": "* enum type",
      "level": 1,
      "required": true,
      "headers": [
        {
          "title": "Members",
          "level": 2,
          "required": true
        }
      ]
    }
  ]
}
```

### CODE REVIEW

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs [hunk 1]
```diff
@@ -75,19 +88,185 @@ public DocumentHeader()
...
+            if (expectedTitle.Equals(foundTitle, comparisonType))
+                return true;
+
+            if (string.IsNullOrEmpty(expectedTitle) || expectedTitle == "*") 
+                return true;
+
+            if (expectedTitle.StartsWith("* ") && foundTitle.EndsWith(expectedTitle.Substring(2), comparisonType)) 
+                return true;
+
+            if (expectedTitle.EndsWith(" *") && foundTitle.StartsWith(expectedTitle[..^2], comparisonType))
```
Comment (baywet) [inline ApiDoctor.Validation/Config/DocumentOutlineFile.cs:114]: add string comparisons for those 3 star comparisons

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs [hunk 2]
```diff
@@ -75,19 +88,185 @@ public DocumentHeader()
...
+                return true;
+
+            if (expectedTitle.EndsWith(" *") && foundTitle.StartsWith(expectedTitle[..^2], comparisonType)) 
+                return true;
+
             return false;
         }
+        internal bool IsMisspelt(DocumentHeader found)
+        {
+            return this.Level == found.Level && this.Title.StringDistance(found.Title) < 5;
```
Comment (baywet) [inline ApiDoctor.Validation/Config/DocumentOutlineFile.cs:121]: why 5?

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs [hunk 3]
```diff
@@ -75,19 +88,185 @@ public DocumentHeader()
...
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType == JsonToken.StartArray)
+            {
+                var jArray = JArray.Load(reader);
+                var allowedHeaders = new List<object>();
+                foreach (var item in jArray)
+                {
+                    bool isConditionalHeader = item.ToString().TryParseJson(out ConditionalHeader conditionalHeader);
```
Comment (baywet) [inline ApiDoctor.Validation/Config/DocumentOutlineFile.cs:219]: this is going to be slow:
1. parse the json to an abitrary model.
2. serialize to string.
3. parse again, with the custom type.

Can't we parse the collection directly?

File: ApiDoctor.Validation/ExtensionMethods.cs [hunk 4]
```diff
@@ -573,7 +573,43 @@ public static bool IsHeaderBlock(this Block block)
                     return false;
             }
         }
+        /// <summary>
+        /// Uses the Damerau-Levenshtein distance algorithm which calculates how different one string is from another 
+        /// </summary>
+        /// <param name="s"></param>
+        /// <param name="t"></param>
+        /// <returns></returns>
+        public static int StringDistance(this string s, string t)
```
Comment (baywet) [inline ApiDoctor.Validation/ExtensionMethods.cs:582]: any reason why we're rolling our own implementation here as opposed to bringing on a dependency?
https://www.nuget.org/packages/Fastenshtein
https://www.nuget.org/packages/FuzzySharp
(and probably many more)

**Review by baywet — CHANGES_REQUESTED** [review]

File: ApiDoctor.Validation/DocFile.cs [hunk 5]
```diff
@@ -675,99 +681,286 @@ protected bool ParseMarkdownBlocks(IssueLogger issues)
...
+        }
+
+        private List<ExpectedHeader> ExtractAndMergeConditionalHeaderArgumentsInList(IReadOnlyList<object> expectedHeaders)
+        {
+            List<ExpectedHeader> allHeaders = new List<ExpectedHeader>();
+            foreach (var header in expectedHeaders)
+            {
+                if (header is ExpectedHeader)
+                {
+                    allHeaders.Add((ExpectedHeader)header);
```
Comment (andrueastman) [inline ApiDoctor.Validation/DocFile.cs:939]:
```suggestion
                if (header is ExpectedHeader expectedHeader)
                {
                    allHeaders.Add(expectedHeader);
```

File: ApiDoctor.Validation/DocFile.cs [hunk 6]
```diff
@@ -675,99 +681,286 @@ protected bool ParseMarkdownBlocks(IssueLogger issues)
...
+                }
+                else if (header is ConditionalHeader)
+                {
+                    var arguments = ExtractAndMergeConditionalHeaderArgumentsInList(((ConditionalHeader)header).Arguments);
```
Comment (andrueastman) [inline ApiDoctor.Validation/DocFile.cs:943]:
```suggestion
                else if (header is ConditionalHeader conditionalHeader)
                {
                    var arguments = ExtractAndMergeConditionalHeaderArgumentsInList(conditionalHeader.Arguments);
```

File: ApiDoctor.Validation/ExtensionMethods.cs (hunk 4 above)
Comment (millicentachieng) [reply ApiDoctor.Validation/ExtensionMethods.cs:582]: Good point. Will use Fastenshtein.

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs (hunk 1 above)
Comment (millicentachieng) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs:114]: Why would that be necessary here?

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs (hunk 2 above)
Comment (millicentachieng) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs:121]: Checking for less distance will give us more accurate information. I'll check for 2 to account for headers that have more than one word.

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs [hunk 7]
```diff
@@ -75,19 +88,185 @@ public DocumentHeader()
...
+        {
+            var header = this.MemberwiseClone() as ExpectedHeader;
+            var childHeaders = new List<object>();
+            foreach (var childHeader in this.ChildHeaders)
+            {
+                if (childHeader is ExpectedHeader expectedChildHeader)
+                    childHeaders.Add(expectedChildHeader.Clone());
+                else
+                    childHeaders.Add(((ConditionalHeader)childHeader).Clone());
+            }
```
[the remark this answers is not in the document]
Comment (millicentachieng) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs:160]: No, we only have the two header types. I've used pattern matching and thrown an exception when any other type is encountered.

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs [hunk 8]
```diff
@@ -75,19 +88,185 @@ public DocumentHeader()
...
+            return this.Level == found.Level && this.Title.StringDistance(found.Title) < 5;
+        }
+
+        public override string ToString()
+        {
+            return this.Title;
+        }
+    }
+
+    public class ExpectedHeader : DocumentHeader
```
[the remark this answers is not in the document]
Comment (millicentachieng) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs:130]: I've used copy constructors instead.

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs [hunk 9]
```diff
@@ -75,19 +88,185 @@ public DocumentHeader()
...
+        public ConditionalHeader Clone()
+        {
+            var header = (ConditionalHeader)this.MemberwiseClone();
+            var arguments = new List<object>();
+            foreach (var arg in this.Arguments)
+            {
+                if (arg is ExpectedHeader headerArg)
+                    arguments.Add(headerArg.Clone());
+                else
+                    arguments.Add(((ConditionalHeader)arg).Clone());
```
[the remark this answers is not in the document]
Comment (millicentachieng) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs]: Refactored to have only one method for copying headers

File: ApiDoctor.Validation/DocFile.cs [hunk 10]
```diff
@@ -694,93 +694,268 @@ protected bool ParseMarkdownBlocks(IssueLogger issues)
...
+                        validationResult != DocumentHeaderValidationResult.OptionalDocumentHeaderMissing)
+                    {
+                        break;
+                    }
+                }
+            }
+            else if (expectedConditionalHeader.Operator == ConditionalOperator.AND)
+            {
+                expectedHeaders[expectedIndex] = expectedConditionalHeader.Arguments.First();
+                expectedHeaders.InsertRange(expectedIndex + 1, expectedConditionalHeader.Arguments.Skip(1));
```
Comment (baywet) [inline ApiDoctor.Validation/DocFile.cs:938]: we should probably check for the collection size/presence of elements before we call first/skip on the arguments.

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs [hunk 11]
```diff
@@ -1,93 +1,177 @@
...
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
         {
-            return this.Level == found.Level && DoTitlesMatch(this.Title, found.Title);
+            if (reader.TokenType == JsonToken.StartArray)
+            {
+                var allowedHeaders = new List<object>();
+                var jArray = JArray.Load(reader);
+                foreach (var item in jArray)
+                {
+                    if (item["condition"] != null)
```
Comment (baywet) [inline ApiDoctor.Validation/Config/DocumentOutlineFile.cs:144]:
```suggestion
                    if (item.ContainsKey("condition"))
```

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs [hunk 12]
```diff
@@ -1,93 +1,177 @@
...
+                    {
+                        var conditionalHeader = item.ToObject<ConditionalDocumentHeader>(serializer);
+                        allowedHeaders.Add(conditionalHeader);
+                    }
+                    else if (item["title"] != null)
```
Comment (baywet) [inline ApiDoctor.Validation/Config/DocumentOutlineFile.cs:149]:
```suggestion
                    else if (item.ContainsKey("title"))
```

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs (hunk 11 above)
Comment (millicentachieng) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs:144]: `item` is of type `JToken` and not dictionary.  Using `ContainsKey` will not work.

Comment (baywet) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs:144]: isn't it a JObject? couldn't we add a type assertion?

Comment (millicentachieng) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs:144]: It's a `JToken`. What are your reservations about performing the check as I am?

Comment (andrueastman) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs:144]: I think @baywet means we can do something like this to verify that the `JToken` is a `JObject`
```cs
if (item is not JObject jobject)
```
Then use the `ContainsKey` from JObject. as indexing may throw an key not found error if I'm not wrong.

**Review by baywet — APPROVED** [review]

File: ApiDoctor.Validation/Config/DocumentOutlineFile.cs (hunk 11 above)
Comment (millicentachieng) [reply ApiDoctor.Validation/Config/DocumentOutlineFile.cs]: Thanks @andrueastman. I've made the update.

**Pull request merged**